src/sema/diagnostics.rs

```diff
@@ -34,17 +34,23 @@ impl Diagnostics {
     }
 
     pub fn push(&mut self, diagnostic: Diagnostic) {
-        if matches!(diagnostic.level, Level::Error) {
-            self.has_error = true;
+        if !self.contents.contains(&diagnostic) {
```

Comment:
This performs a search trough the entire content of the vector. We push diagnostics many times throughout the code. If you think it is necessary checking for repeated diagnostics, I suggest using a hash table.